lib/routes/shared/security_pin/prompt_pin_code.dart

```diff
@@ -0,0 +1,178 @@
+  final Widget route;
+
+  LockScreen({Key key, this.label, this.dismissible = false, this.changePin = false, this.route}) : super(key: key);
+
+  @override
+  _LockScreenState createState() => new _LockScreenState();
+}
+
+class _LockScreenState extends State<LockScreen> {
+  UserProfileBloc _userProfileBloc;
+  BreezUserModel _user;
+  String _label;
+
+  String _enteredPinCode = "";
+  String _errorMessage = "";
+  String _tmpPinCode = "";
+
+  bool _isInit = false;
+
+  @override
+  void initState() {
+    super.initState();
+    _setLabel(widget.label ?? "Enter your PIN");
+  }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    if (!_isInit) {
+      _userProfileBloc = AppBlocsProvider.of<UserProfileBloc>(context);
+      _userProfileBloc.userStream.listen((user) {
+        _user = user;
+      });
+      _isInit = true;
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return WillPopScope(
+      onWillPop: () => Future.value(false),
+      child: Scaffold(
+        appBar: widget.dismissible
+            ? new AppBar(
+                iconTheme: theme.appBarIconTheme,
+                textTheme: theme.appBarTextTheme,
+                backgroundColor: theme.BreezColors.blue[500],
+                leading: backBtn.BackButton(
+                  onPressed: () {
+                    Navigator.pop(context, false);
+                  },
+                ),
+                elevation: 0.0,
+              )
+            : null,
+        body: PinCodeWidget(
+          _label,
+          _enteredPinCode,
+          widget.dismissible,
+          _errorMessage,
+          (numberText) => _onNumButtonPressed(numberText),
+          (enteredPinCode) => _setPinCodeInput(enteredPinCode),
+        ),
+      ),
+    );
+  }
+
+  _onNumButtonPressed(String numberText) {
+    _setErrorMessage("");
+    if (_enteredPinCode.length < PIN_CODE_LENGTH) {
+      _setPinCodeInput(_enteredPinCode + numberText);
+    }
+    if (_enteredPinCode.length == PIN_CODE_LENGTH) {
+      if (widget.changePin) {
+        _handleChangePin();
+      } else {
+        _handleUnlockPin();
+      }
+    }
+  }
+
+  _handleChangePin() {
+    if (_tmpPinCode.isEmpty) {
+      // If a new PIN is being set prompt user to enter the PIN again
+      Future.delayed(Duration(milliseconds: 300), () => _setPinCodeForValidation()); // Wait 300ms for scale animation to end
+    } else {
+      // Check if PINs match if a new PIN is being set or current PIN is being changed
+      Future.delayed(Duration(milliseconds: 300), () => _matchPinCodes(_enteredPinCode == _tmpPinCode));
+    }
+  }
+
+  _setPinCodeForValidation() {
+    setState(() {
+      _tmpPinCode = _enteredPinCode;
+    });
+    _setLabel("Re-enter your new PIN");
+    _setPinCodeInput("");
+  }
+
+  void _matchPinCodes(bool isMatched) {
+    if (isMatched) {
+      _setPinCode(_enteredPinCode);
+    } else {
+      _setLabel("Enter your new PIN");
```

Comment:
In case the pin code doesn't match shouldn't we revert to the first state and clear the `_tmpPinCode`?
Currently the `_tmpPinCode ` contains the old value

Reply:
[Fixed](https://github.com/breez/breezmobile/pull/128/commits/1471db5e2a95d320271a0deb1e2fb8c734541e84)